sdk/core/azure-core/tests/test_serialization.py

```diff
@@ -503,12 +505,682 @@ def _flattened_model_assertions(model):
+        stock_count: int = rest_field(name="stock_count")  # Underscore in both
+
+    class MsrestProduct(MsrestModel):
+        _attribute_map = {
+            "product_id": {"key": "productId", "type": "str"},
+            "product_name": {"key": "ProductName", "type": "str"},  # Casing difference
+            "unit_price": {"key": "unit-price", "type": "float"},  # Dash in REST name
+            "stock_count": {"key": "stock_count", "type": "int"},  # Underscore in both
+        }
+
+    def _test(result):
+        assert result["product_id"] == "p123"
+        assert result["product_name"] == "Widget"
+        assert result["unit_price"] == 19.99
+        assert result["stock_count"] == 42
+
+    hybrid_product = HybridProduct(product_id="p123", product_name="Widget", unit_price=19.99, stock_count=42)
+    assert hybrid_product["productId"] == "p123"
+    assert hybrid_product["ProductName"] == "Widget"
+    assert hybrid_product["unit-price"] == 19.99
+    assert hybrid_product["stock_count"] == 42
+    _test(as_attribute_dict(hybrid_product))
+
+    msrest_product = MsrestProduct(product_id="p123", product_name="Widget", unit_price=19.99, stock_count=42)
+    _test(as_attribute_dict(msrest_product))
+
+
+def test_as_attribute_dict_datetime_serialization():
+    class HybridEvent(HybridModel):
+        event_id: str = rest_field(name="eventId")
+        start_time: datetime = rest_field(name="startTime")
+        end_time: datetime = rest_field(name="endTime")
+        created_date: date = rest_field(name="createdDate", format="date")
+        reminder_time: time = rest_field(name="reminderTime", format="time")
+        duration: timedelta = rest_field(format="duration")
+
+    class MsrestEvent(MsrestModel):
+        _attribute_map = {
+            "event_id": {"key": "eventId", "type": "str"},
+            "start_time": {"key": "startTime", "type": "iso-8601"},
+            "end_time": {"key": "endTime", "type": "iso-8601"},
+            "created_date": {"key": "createdDate", "type": "date"},
+            "reminder_time": {"key": "reminderTime", "type": "time"},
+            "duration": {"key": "duration", "type": "duration"},
+        }
+
+        def __init__(
+            self,
+            *,
+            event_id: str,
+            start_time: datetime,
+            end_time: datetime,
+            created_date: date,
+            reminder_time: time,
+            duration: timedelta,
+            **kwargs
+        ):
+            super().__init__(**kwargs)
+            self.event_id = event_id
+            self.start_time = start_time
+            self.end_time = end_time
+            self.created_date = created_date
+            self.reminder_time = reminder_time
+            self.duration = duration
+
+    def _test(result):
+        assert result["event_id"] == "e789"
+        assert isinstance(result["start_time"], str)
+        assert result["start_time"] == "2023-05-15T09:00:00Z"
```

Comment:
todo: there's some sigfig differences between hybrid and msrest datetime serialization. Goign to take a look tomorrow